### ISSUE TITLE
Document environments/architecture in relation to pricing

### ISSUE THREAD

**Issue by asim** [m3o/dev#537]
Environments are a concept baked into Micro 3.0 that define separate deployments of Micro and enable the separation of local, staging and production which we refer to as local, dev, platform. Namespaces are separately a form of multitenancy within an environment that can be used to separate tenants, products or feature development.

There has been some discussion about environments in relation to our pricing and whether they should in fact be namespaces within a single cluster. I will let @chrusty chime in here based on historic experience but this poses major issues when it comes to scaling and security. Architecturally separate clusters are superior for separation of concerns while a "meta" cluster can be used as a control plane to orchestrate everything as multiple clusters become cumbersome to manage.

Kubernetes supports federation and this echoes these sentiments. See the doc below.

https://platform9.com/blog/kubernetes-federation-what-it-is-and-how-to-set-it-up/#:~:text=Kubernetes%20Federation%20takes%20this%20concept,are%20members%20of%20the%20federation.

For our requirements we are defining an experience much like you'd see within an existing tech company. Which always separate infrastructure for different environments including spinup of isolated environments for branch development or PRs. Our use of local, dev, platform is to instill best practices but also separate what is random hacking or small projects from a supported scalable production environment. It may seem "cool" or innovative to use a different approach like a single large cluster but this inevitably leads us down the same path of splitting the clusters.

Node pools are one option that has been discussed for this separation in a single cluster but it still makes use of the same kubernetes API and any form of overloading here tends to cripple a cluster. Node pools are otherwise useful to separate different types of workloads within the same environment so that high CPU services go to one place while high mem usage go elsewhere. It also enables colocating services which may be written and grouped together by one team e.g payments services, location services, etc.

This issue is the start of documenting the rationale behind environments and their separation in relation to pricing, the related isolation and evolution based on federation. Federation as a concept will likely find it's way into Micro when the time comes.

Pricing will evolve and Auth may also evolve to oauth so a single account can be used across many environments but we must also consider that environments are still isolated separate micro installations so oauth won't be mandatory but one of the forms of Auth possible. Accounts per env will still exist.

From a product perspective we offer
- Dev as a free environment in the cloud
- Platform as a paid and supported environment with SLAs
- Cloud as a "contact us" for isolated environments managed or self managed

We should promote a Local, Dev, Platform, Cloud product flow and pricing model which for both technical and business purposes scale with different user requirements over time.

1. Run OSS locally (individual)
2. Push to Dev for cloud hosted small projects (3-5 people)
3. Move to Platform for supported production workloads (8-10 people)
4. Switch to isolated Cloud environments as your organisation scales (25+ people)

**Comment by chrusty** [comment]
As long as our automation is on point then the overhead of running more clusters isn't such a big deal. In fact, my opinion is that wherever we have the opportunity to naturally segregate resources for free then we should take it. The key considerations (as I see them) in our circumstance are:

- Maintain the least amount of compute overhead as possible by employing horitontal pod autoscaling and cluster autoscaling. Having several clusters at 1% utilisation would be a crime against clouds.
- One thing we would really miss out on by having several clusters is the ability for a customer to "upgrade" from the free tier to the paid tier and have their services / configs remain in place. Some crufting can get us around this anyway I'm sure, but it should be noted. I can't quite remember, was either @arbarlow or @ben-toogood who brought this up initially.
- The most important aspect for me is that we _do_ require a cluster where we can test things safe in the knowledge that we're not going to break anyone's service (whether free OR paid). I'd consider this as a fundamental requirement for us, regardless of expense.

**Comment by asim** [comment]
I'll write all this up in a design doc as I don't think it exists. We need to accept the state of the current working solution and if there's any issues or concerns interms of utilisation or upgrades across a couple clusters then we design and understand that but based on customer feedback rather than just doing it. It might be a non problem. Also yes I stole staging for Dev. I think in our case determining whether a long live staging is needed or whether that can be something ephemeral.

The question of multi-environment per cluster is interesting but I'll need to see more as a full on design doc to understand what means. Tbh I'm not sure it's worth the pain. 

In terms of pricing tiers related to envs. This might seem strange because most companies do that in a single production cluster but given environments directly relate to this concept and the pricing is done in that way this makes sense. It's more likely the pricing evolves rather than the architecture and you get access to multiple environments. Micro builds in that cross env switching well. The multiple accounts, upgrades, etc is a separate piece to be considered.

**Comment by asim** [comment]
@domwong can you chime in on this before I start writing a doc. Over the long term I think ultimately you'll own this sort of decision.

**Comment by arbarlow** [comment]
So my 2p on this is that for a user, they'll want one signup, one account, many projects or namespaces.

There is no AWS dev, or GCP dev for this reason, it's internal to them and I feel for our testing we should also do the same, a tiny cluster for that purpose.

I agree with @chrusty in that, absolutely key to all of this is security and resource isolation, but that's absolutely key anyway. Any cluster, paid or not, cannot have any holes in that model, or we're screwed. So we *must* solve that problem first architecturally and then work from there, I feel atm we're making a tech decision based on restrictions that's effecting the user experience and confusing new users.

I think we should have a chat or similar to talk about resource isolation and get that rock solid. but if we can solve those problems I think my approach would be..

1 m3o cluster (responsible for all signup, accounts, cluster management)
n sharded clusters (the more we grow, the more we add, limiting the amount of customers per shard)

This would mean that the problem becomes managing many clusters, which is fine (we just talk to the right cluster at the right time, we may also need a load balancing layer), but from a users' perspective, it feels like one platform.

In terms of their upgrade/env experience, I would imagine something like the following?

Unlimited namespaces, unlimited users per namespace.

Per namespace:
- (Optional) 200 GBP per month silver support level, 1h tickets
- (Optional) 600 GBP per month gold support level, managed, 1h tickets, private shared slack channel
- 10 GBP per 1 unit of compute per month (billed per 100m, minimum 200m)
- 1 GBP per 1 GB of memory per month (billed per 1GB, minimum 1 GB)
- xx GBP per GB disk space per month
- xx per 1000 events
- xx per 1m traces

This would upgrading is just adding packages or features and paying.

TL;DR what we need to solve is isolation for everyone

**Comment by domwong** [comment]
Only just seen this.

My 2p 

The current approach of separate dev and platform tier clusters is not the long term solution. My current thinking is if we move to the sharding / multiple clusters managed by a single "meta" cluster where our "prod" environment lives I think that's the long term solution. Moving to this model doesn't even have to involve a painful migration if we do it the right way. 

Remember, we only did it like this to validate the theory that a free tier is fundamental to gaining traction. I think it's proving/proved that point but we mustn't be afraid to fix the bad architectural decisions; the earlier we do these things the better, but like I said it's entirely possible to do it in a non painful way. 

Fixing this would enable a much more robust implementation for things like `local -> dev -> prod` and upgrading an existing user's free tier to paid without needing them to move everything to prod.

**Comment by asim** [comment]
I'll turn this into some sort of related doc here in github along with the comments. It seems we'll end up down the "meta" control plane route when the time comes however actually designing upfront will be fairly important and will likely require an "environments" service to do the management of that which ties into our Cloud tier offering and the potential for automation of that along with ephemeral testing environments.

**Comment by domwong** [comment]
nobody touches code until we see a design
